### PR TITLE
tests: determine runners dynamically based on formulae to be tested

### PR DESCRIPTION
**Pull request by carlocab**
Running the `test_deps` job can be rather expensive for formulae which
have no dependents. Let's skip those.

This cribs heavily from the `dispatch-rebottle` workflow.

There are also edge cases where it doesn't get things quite right. For
example, `brew determine-dependent-runners` incorrectly concludes that
dependents must be tested if a PR modifies two formulae where one
depends on the other, but the two formulae otherwise have no dependents.

Still, this should be an improvement on the current situation.

### CODE REVIEW

**Review by p-linnane — DISMISSED** [review]

File: cmd/determine-dependent-runners.rb
Comment (carlocab) [inline cmd/determine-dependent-runners.rb:1]: This logic could probably be extended to handling generating the required runners for the `tests` job too. The logic there is likely even simpler.

**Comment by carlocab** [comment]
Ok, I've updated this to dynamically determine the required runners for both `tests` and `test_deps` jobs based on `testing_formulae` and `deleted_formulae`.

Benefits of doing this:
1. We only request Linux runners when testing Linux-only formulae (or dependents), and only macOS runners when testing macOS-only formulae (or dependents).
2. We skip requesting any runners at all if there are no dependents to test after having accounted for `testing_formulae`. i.e. Some `testing_formulae` may have dependents, but if those dependents are already also in `testing_formulae` then there are no more dependents to test.

As an example of how this process works, I've rev bumped a Linux-only formulae. We can see that we only request Linux runners for both `tests` and `test_deps`.

File: cmd/determine-test-runners.rb [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  end
+end
+
+module Homebrew
+  module_function
+
+  def determine_test_runners_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `determine-test-runners` <testing-formulae> [<deleted-formulae>]
```
Comment (MikeMcQuaid) [inline cmd/determine-test-runners.rb:51]: Sorry, I don't want to stall progress too hard here but: this is the sort of logic I feel uncomfortable being in Homebrew/homebrew-core where it can't really be tested. Ideally it'd be in `brew test-bot` or Homebrew/brew and have some tests written to exercise it.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -31,6 +31,18 @@ jobs:
...
       HOMEBREW_SIMULATE_MACOS_ON_LINUX: 1
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - run: brew test-bot --only-tap-syntax
+
+  formulae_detect:
+    if: github.repository == 'Homebrew/homebrew-core' && github.event_name == 'pull_request'
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:42]: Note that I think we should think sooner rather than later about having the `merge_group` job run `brew fetch` for these formulae that are changed so I think it's worth continuing to run this.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -213,26 +222,50 @@ jobs:
...
+      runners: ${{steps.determine-runners.outputs.runners}}
+      runners_present: ${{steps.determine-runners.outputs.runners_present}}
+      dependent_runners: ${{steps.determine-dependent-runners.outputs.runners}}
+      dependent_runners_present: ${{steps.determine-dependent-runners.outputs.runners_present}}
+    env:
+      HOMEBREW_LINUX_RUNNER: ${{needs.setup_tests.outputs.linux-runner}}
+      HOMEBREW_LINUX_CONTAINER_IMAGE: ghcr.io/homebrew/ubuntu22.04:master
+      HOMEBREW_LINUX_CONTAINER_OPTIONS: --user=linuxbrew -e GITHUB_ACTIONS_HOMEBREW_SELF_HOSTED
+      HOMEBREW_LINUX_WORKDIR: /github/home
+      HOMEBREW_LINUX_TIMEOUT: 4320
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:243]: I'd just hardcode this in wherever the Ruby script ends up living.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -213,26 +222,50 @@ jobs:
...
-    if: ${{github.event_name == 'pull_request' && fromJson(needs.setup_tests.outputs.syntax-only) == false}}
+    needs: [tap_syntax, setup_tests, setup_runners]
+    if: >
+      github.event_name == 'pull_request' &&
+      fromJson(needs.setup_tests.outputs.syntax-only) == false &&
+      fromJson(needs.setup_runners.outputs.runners_present)
     strategy:
       matrix:
-        include:
-          - runner: "13-arm64-${{github.run_id}}-${{github.run_attempt}}"
```
Comment (MikeMcQuaid) [inline .github/workflows/tests.yml:222]: This approach won't work if we ever want to make a 🟢 `tests` build a requirement because you'll need to have the named build actually run in that case.

**Review by MikeMcQuaid — COMMENTED** [review]
Nice work so far! Need to move most of this Ruby logic elsewhere but I like the direction this is taking!

File: .github/workflows/tests.yml (hunk 4 above)
Comment (carlocab) [reply .github/workflows/tests.yml:222]: Yup, I mention this at https://github.com/Homebrew/homebrew-core/pull/127236#issuecomment-1492993923.

> If we want to make `tests` a required status check at some point, it should be straightforward to modify this to run a job on an `ubuntu-latest` runner that does nothing.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: @carlocab Gotcha. I think I'd like that to be part of the implementation before it's merged, if that's alright 🙇🏻

File: cmd/determine-test-runners.rb (hunk 1 above)
Comment (carlocab) [reply cmd/determine-test-runners.rb:51]: Nothing stops us from adding tests for this right here, though 🙂

I can add some, if you like.

Comment (MikeMcQuaid) [reply cmd/determine-test-runners.rb:51]: @carlocab None of the other `cmd/` in homebrew-core have meaningful tests. It's nicer for the tests and logic to live elsewhere (`test-bot` or `brew`) because they have unit tests suites, Sorbet, etc. setup.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (carlocab) [reply .github/workflows/tests.yml:46]: I've added a `brew test-bot --only-formulae-detect` step back to `tap_syntax`.

File: cmd/determine-test-runners.rb (hunk 1 above)
Comment (carlocab) [reply cmd/determine-test-runners.rb:51]: Ok, let me look at moving this to `test-bot` or `brew`.

File: .github/workflows/tests.yml (hunk 4 above)
Comment (carlocab) [reply .github/workflows/tests.yml:222]: Implemented.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: You can see the latest run deploy an `ubuntu-latest` runner that does basically nothing. (i.e. a `mkdir` here, an `rm -rvf` there...)

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: Thanks @carlocab 🙇🏻

File: .github/workflows/tests.yml (hunk 2 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:46]: @carlocab Thanks 🙇🏻

File: .github/workflows/tests.yml (hunk 4 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: @carlocab Actually, I don't think this quite works as expected/needed:
- the name `tests (ubuntu-latest, true)` seems to have changed. Not opposed to a change here but I think getting rid of that `true` would be preferable.
- this doesn't seem to be creating equivalently named jobs for e.g. macOS which means we won't be able to prevent e.g. someone merging a build when macOS is failing.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: > * the name `tests (ubuntu-latest, true)` seems to have changed. Not opposed to a change here but I think getting rid of that `true` would be preferable.

This is a GitHub thing -- the `true` is from the second part of the `matrix` job. Not quite sure how to get GitHub to stop showing parts of the build matrix, but I'll look into it.

> * this doesn't seem to be creating equivalently named jobs for e.g. macOS which means we won't be able to prevent e.g. someone merging a build when macOS is failing.

Why would it create a macOS job? There are no formulae to test. If there were formulae to be tested here that supported macOS, then macOS jobs would be deployed.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: > If we don't create some sort of macOS job or job-named-the-same-as-macOS-job: we can never require the macOS job be 🟢 before merge.

I'm confused here -- isn't the point to require 🟢 status for `tests` before merging? If we do that, then failing formulae builds cannot be merged, even after this change.

> How much time does a change like this PR save and how many formulae does this affect?

At least 5 minutes per PR for formulae with no dependents (which is most of them). Plus the extra time saved from waiting less time in the queue because they're waiting for runners spending 4~ minutes doing a no-op job, which happens to pretty much every formula. So somewhere between 4-9 minutes per PR, I think. That's also not counting the time we waste re-running jobs because they timed out because of the extra time they spent on no-op stuff.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: Looks like the way to do required checks for `matrix` jobs is to add a check for jobs after them that depend on the `matrix` jobs: https://brunoscheufler.com/blog/2022-04-09-the-required-github-status-check-that-wasnt

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: > status for the `tests` job before merging

There is no `tests` job. When you use a `matrix`: each part of the matrix is a separate job. You need to require success of each of these with the same name. If we ever want to prevent people from merging a e.g. failing macOS 13 ARM job: we always need to generate a job with that name and it always needs to be 🟢.

> At least 5 minutes per PR for formulae with no dependents (which is most of them).

To be clear: I think it's reasonable to skip the `test_deps` job in these situations: I just don't think we ever want to skip any `tests` jobs. Eventually: `CI-syntax-only` should will need to go away.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: > Looks like the way to do required checks for `matrix` jobs is to add a check for jobs after them that depend on the `matrix` jobs

Yes, if you must have a dynamic metric. I don't really like that, though; it means if the matrix logic ends up not working as you expect: you don't actually have the protections you think you do.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: > I don't really like that, though; it means if the matrix logic ends up not working as you expect: you don't actually have the protections you think you do.

I am very confident about the logic here for the `tests` job. (On the other hand, I am less confident about it for `test_deps`, but edge cases should be relatively rare. This is because `HOMEBREW_SIMULATE_MACOS_ON_LINUX` doesn't let you specify which macOS version you want to simulate.)

In any case: we might need to make a post-`tests` job as the required check even without this. We still have some `tests` jobs that take 24+ hours, and when that happens, failures are inevitable because we can't stop `GITHUB_TOKEN` from expiring.

> Eventually: `CI-syntax-only` should will need to go away.

Being able to skip OSes in `tests` would go a long way towards this, I think.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: > I am very confident about the logic here for the tests job.

Ok. As long as much of this logic as possible is pushed to Homebrew/brew and/or Homebrew/homebrew-test-bot and has tests over there: I'm fine with this 👍🏻 

> failures are inevitable because we can't stop `GITHUB_TOKEN` from expiring.

I wouldn't say inevitable here. This seems exactly the sort of thing that we need to fix at some point.

> Being able to skip OSes in tests would go a long way towards this, I think.

Agreed.

Comment (carlocab) [reply .github/workflows/tests.yml:222]: Is the fake "macOS 13-arm64" runner in the latest CI run here more like what you're looking for?

It's fake because it's actually `ubuntu-latest`.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:222]: I think your `conclusion` approach would be fine here. It has just occurred to me that even the current format (`CI / tests (13-arm64-4626220015-1) (pull_request)`) we can't actually require because the test name changes every time 😬

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -40,18 +56,44 @@ jobs:
         id: set-up-homebrew
         uses: Homebrew/actions/setup-homebrew@master
 
-      - run: brew test-bot --only-tap-syntax
-
       - run: brew test-bot --only-formulae-detect
-        if: github.event_name == 'pull_request'
         id: formulae-detect
 
+      - name: Fetch detected formulae bottles
```
Comment (carlocab) [inline .github/workflows/tests.yml:62]: @MikeMcQuaid I've added a test to make sure we can fetch the just-built bottles before merging. We can do something like
```yaml
if: github.event_name == 'merge_group'
```
so it's run just before an attempt to merge.

And, yes, don't worry, I'll look at moving this to `brew`/`test-bot`/etc. too. Just writing it down here as a quick POC.

Comment (carlocab) [reply .github/workflows/tests.yml:65]: Homebrew/homebrew-test-bot#902

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:67]: @carlocab That's really great, thanks!

File: .github/workflows/tests.yml [hunk 6]
```diff
@@ -356,28 +391,92 @@ jobs:
           brew test-bot --only-cleanup-after
           rm -rvf bottles
 
-  test_deps:
-    needs: [tap_syntax, setup_tests, tests]
+  setup_dep_runners:
```
Comment (carlocab) [inline .github/workflows/tests.yml:394]: I made this a separate job since it takes much longer than determining runners for the `tests` job, so I'd rather not end up having the `tests` job wait for it when the `tests` job doesn't need it.

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -356,28 +391,92 @@ jobs:
...
+          printf '::error ::Unexpected outcome!\n'
+          printf '::error ::`tests` job result: %s\n' "$result" # success/skipped
+          printf '::error ::runners assigned:   %s\n' "$runners_present" # true/false
+          printf '::error ::syntax-only:        %s\n' "$syntax_only" # true/false
+          exit 1
+
+  test_deps:
+    needs: [tap_syntax, setup_tests, setup_dep_runners, tests]
+    if: >
+      (success() || failure()) &&
```
Comment (carlocab) [inline .github/workflows/tests.yml:470]: I might get some pushback here: but this just restores the original behaviour of testing dependents before we split it into a separate job. (i.e. it is run even when the `tests` job fails)

**Comment by carlocab** [comment]
Everything working as expected here.

With https://github.com/Homebrew/homebrew-core/labels/CI-published-bottle-commits: https://github.com/Homebrew/homebrew-core/actions/runs/4658816607/jobs/8244943700?pr=127236#step:5:38

Without https://github.com/Homebrew/homebrew-core/labels/CI-published-bottle-commits: https://github.com/Homebrew/homebrew-core/actions/runs/4658626863 (tests cancelled by force push)

**Comment by carlocab** [comment]
As a bonus: all our workflows will pass `actionlint` once this is merged.

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -356,28 +391,92 @@ jobs:
...
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - name: Determine runners to use for test_deps job
+        id: determine-dependent-runners
+        run: brew determine-test-runners --dependents --eval-all "$TESTING_FORMULAE"
+
+  conclusion:
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:419]: Could some of this logic too also be moved into another command?

**Review by MikeMcQuaid — APPROVED** [review]

File: .github/workflows/tests.yml (hunk 7 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:470]: @carlocab Yeh, I'd rather we didn't do this please. It wastes runner time on jobs we're not going to ship.

File: .github/workflows/tests.yml [hunk 9]
```diff
@@ -31,6 +31,18 @@ jobs:
...
     env:
       HOMEBREW_SIMULATE_MACOS_ON_LINUX: 1
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - run: brew test-bot --only-tap-syntax
+
+  formulae_detect:
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:41]: Pretty sure if it's not a required job: it won't block merge queue merges (and perhaps not even get run).

[the remark this answers is not in the document]
Comment (carlocab) [reply .github/workflows/tests.yml:41]: Ok, let me know when you're around to make it required and I'll merge this then.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (carlocab) [reply .github/workflows/tests.yml]: It really is still needed for some PRs. As a compromise: I've made it so that `test_deps` is run despite failing `tests` only when https://github.com/Homebrew/homebrew-core/labels/CI-no-fail-fast is set.

File: .github/workflows/tests.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:419]: @carlocab yup, fine to merge-as is for this bit.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml]: > It really is still needed for some PRs.

Which ones and why? These are the sort of jobs that are going to prevent our ability to make `tests` a required CI status.


> As a compromise: I've made it so that test_deps is run despite failing tests only when [CI-no-fail-fast](https://github.com/Homebrew/homebrew-core/labels/CI-no-fail-fast) Continue CI tests despite failing GitHub Actions matrix builds. is set.

Works for now, happy to merge as-is here 👍🏻

**Review by MikeMcQuaid — APPROVED** [review]
Some suggested changes for future PRs but happy to land this one as-is. Great work @carlocab!

**Comment by MikeMcQuaid** [comment]
> CI / ${{ matrix.name }} (pull_request) Skipped
> CI / ${{ matrix.name }} (deps) (pull_request) Skipped

This is pretty ugly, I'd love to see us avoid this eventually.

File: .github/workflows/tests.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (carlocab) [reply .github/workflows/tests.yml:41]: Queued. Will need to make formulae_detect required soon.

**Comment by carlocab** [comment]
Unfortunately fetching bottles isn't ready for `merge_group` yet: #128054

We can still run it once bottles are pushed to the PR, though.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (carlocab) [reply .github/workflows/tests.yml]: > Which ones and why?

Typically: audit failures or `GITHUB_TOKEN` expiry errors. The most common type of audit failure we end up ignoring is an online check failure (e.g. `brew audit` reports that the source tarball is not available but `brew install` and `brew fetch` seemed to have no problem with it -- bizarre, but it happens). Less commonly, re-running CI just takes too long (several hours) that it's just a better use of resources to fix the audit failure in a follow-up PR.

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml]: > e.g. brew audit reports that the source tarball is not available

IMO we should just remove all these checks considering they are basically duplicates.

> GITHUB_TOKEN expiry errors

I think the new jobs should address this? Alternatively, using a BrewTestBot PAT?

Comment (MikeMcQuaid) [reply .github/workflows/tests.yml]: Alternatively: we just pull literally all the `brew audit` runs into a `tap_syntax` like job that's Linux-only, run on the detected formulae and run before and `tests` jobs even run. Easily rerun in that case, too.

Comment (carlocab) [reply .github/workflows/tests.yml]: > IMO we should just remove all these checks considering they are basically duplicates.

Yea, 👍 on that. The only difficult part is that we have _a lot_ of audits, so it's a bit painful to go through them all to weed out the ones we don't want anymore.

> I think the new jobs should address this? Alternatively, using a BrewTestBot PAT?

New jobs will help. BrewTestBot PAT will not -- no secrets available on PR workflows.

> Alternatively: we just pull literally all the `brew audit` runs into a `tap_syntax` like job that's Linux-only, run on the detected formulae and run before and `tests` jobs even run. Easily rerun in that case, too.

I've been thinking this too. This might be the nicest solution.
